Mirek/split options2

mirefek opened the pull request.

LasseBlaauwbroek commented on graph2tac/loader/py_data_server.py:
```diff
@@ -171,23 +171,67 @@ def cluster_to_graph(self, cluster):
                  data_dir: Path,
                  max_subgraph_size,
-                 split: tuple[int, int, int] = (8, 1, 1),
+                 split = (8, 1, 1),
```
Isn't it time for us to get rid of the concept of holdout-set? I think we have all agreed in the past that we will almost certainly not use this.

mirefek replied:
Up to discussion. I don't mind too much allowing more than two split labels.

LasseBlaauwbroek replied:
I would at least change the default to `(9, 1, 0)`.

mirefek commented:
https://github.com/IBM/graph2tac/pull/100 Merged without issues

mirefek commented on graph2tac/loader/py_data_server.py:
```diff
@@ -171,23 +171,67 @@ def cluster_to_graph(self, cluster):
+        self.data_server = None # to be set by data_server
+        self.prefixes_per_label = []
+        for label, prefixes in enumerate(prefixes_per_label):
+            if isinstance(prefixes, str):
```
I just want the config being flexible, where we parse the config is up to discussion

LasseBlaauwbroek replied:
In my opinion, you should always determine the shape or your data as early as possible. So I'd do it when parsing the yaml file.

mirefek commented on graph2tac/loader/py_data_server.py:
```diff
@@ -171,23 +171,67 @@ def cluster_to_graph(self, cluster):
+                prefixes = list(prefixes)
+            self.prefixes_per_label.append((label+1, prefixes))
+    def datapoint(self, d : Definition):
+        fname = self.data_server.graphid_to_fname[d.node.graph]
```
In case there would be another Splitter taking something else from the data_server

mirefek commented on graph2tac/loader/py_data_server.py:
```diff
@@ -171,23 +171,67 @@ def cluster_to_graph(self, cluster):
-        self.split_random_seed = split_random_seed
         self.restrict_to_spine = restrict_to_spine
 
+        if isinstance(split, tuple) and isinstance(split[0], int):
+            self.split = SplitByHash(split, split_random_seed)
+        elif isinstance(split, tuple) and isinstance(split[0], (str, list)) and isinstance(split[0][0], str):
+            self.split = SplitByFilePrefix(split)
+        elif isinstance(split, Splitter):
+            self.split = split # ignore split_random_seed
+        else:
+            raise Exception("Cannot interpret split:", split)
+        self.split.data_server = self
```
Passing directly a Splitter is among the options. I agree I am sort of lazy / hacky here with keeping backward compatibility. At some point, we need to parse the config file, and create the appropriate Splitter. Now, I am doing it here but I agree there could be a better option.

mirefek replied:
For example, TF2 is using the old split argument, and I am not sure if we are already fully getting rid of it.

LasseBlaauwbroek replied:
Yes, we are getting rid of TF2 (Jason should be working on it). I think that the time for keeping backwards compatibility is over. I would just change it to what we think is good code.

jasonrute replied:
See my suggestion below.  And yes, I'm getting rid of TF2 this week hopefully.

jasonrute replied:
Actually, I'll just repeat it.  We can use this pattern often used by Fidel in the yaml files:
```
split_method: hash
split:
  proportions:
  - 0
  - 1
  split_random_seed: 0
```
So then we pass that to the DataServer as `split_method: str` and `split: dict`.  This will make it easy to add more split methods as needed.

LasseBlaauwbroek commented:
As a security measure, I think it might be a good idea to put in a check that all of the prefixes are self-consistent. That is, none of the files in the training set should reference the validation set. Otherwise we should error out.

mirefek commented:
Good idea, I will implement some checking

LasseBlaauwbroek commented on graph2tac/loader/data_server.py:
```diff
@@ -171,23 +171,67 @@ def cluster_to_graph(self, cluster):
+        ident_64 = np.array(d.node.identity, dtype = np.uint64).item()
+        return get_split_label(ident_64, self.split, self.random_seed)
+    def cluster(self, d : list[Definition]):
+        return 0
```
Why are clusters always part of training?

mirefek replied:
That is the old behavior

LasseBlaauwbroek replied:
Was there an actual historical reason for this, or is this just a bug? If the latter, I'd fix it...

mirefek replied:
to me, it is a logical behavior

mirefek replied:
when you randomly split on lemmata

mirefek replied:
what other split on definitions do you imagine?

LasseBlaauwbroek replied:
Ah, okay. I misunderstood the purpose of `datapoint` and `cluster`. I get what is happening now, but I was interpreting them as doing exactly the same, except that one took a list. Perhaps rename them to `lemma` and `definition_cluster`?

mirefek replied:
I wouldn't `split.lemma_split` feels having unnecessary repetition to me.

jasonrute replied:
That is a fine point.  We can leave it as `lemma` and `definition cluster`.

LasseBlaauwbroek commented:
Question: Currently you are keeping a mapping from graphid to filenames. Then, for every datapoint, you lookup its graphid, translate to the filename and try to decide in which split the file is. Wouldn't it be simpler and faster, to pre-calculate for each graphid to which split it belongs? Then, deciding where a datapoint belongs is a question of one map-lookup.

mirefek commented:
Good point, I originally went for the easy fix, I will rewrite it this way.

jasonrute commented on graph2tac/tfgnn/dataset.py:
```diff
@@ -493,16 +477,16 @@ class DataServerDataset(Dataset):
-    def __init__(self, data_dir: Path, max_subgraph_size: int = 1024, **kwargs):
+    def __init__(self, data_dir: Path, split = None, split_random_seed = 0, max_subgraph_size: int = 1024, **kwargs):
         """
         @param data_dir: the directory containing the data
         @param max_subgraph_size: the maximum size of the returned sub-graphs
         @param kwargs: additional keyword arguments are passed on to the parent class
         """
         self.data_server = DataServer(data_dir=data_dir,
                                       max_subgraph_size=max_subgraph_size,
-                                      split=(1,0,0),
-                                      split_random_seed=0)
+                                      split = tuple(split),
+        )
```
But also see my proposal where we package the random seed inside the split.

jasonrute commented on graph2tac/tfgnn/dataset.py:
```diff
@@ -493,16 +477,16 @@ class DataServerDataset(Dataset):
     definition_data_spec = (loader_graph_spec, num_definitions_spec, definition_names_spec)
 
-    def __init__(self, data_dir: Path, max_subgraph_size: int = 1024, **kwargs):
+    def __init__(self, data_dir: Path, split = None, split_random_seed = 0, max_subgraph_size: int = 1024, **kwargs):
```
Again, I would removed the defaults.  In particular, `split = None` would likely crash the code, right???

jasonrute commented:
Also, when you merge in main, which now has tests, please update the tests (both the parameter files and results) and add a test for splitting on prefix.  You can add the test in one of the datasets.  (I want to see if my README in `tests/README.md` is good, so first look there to see if you can figure out how to do all this with the tests.  If not, I'm happy to help and we can update the README.)

mirefek commented on graph2tac/tfgnn/dataset.py:
```diff
@@ -172,25 +158,23 @@ def graph_constants(self) -> GraphConstants:
         Compute statistics for the proof-state and definition datasets.
 
         @param split: a pair of (not necessarily normalized) probabilities for the train and validation splits
         @param split_random_seed: the seed to use for the train/validation split
```
What is wrong?

jasonrute replied:
The doc string mentions `@param split_random_seed`, `@param split_random_seed` which are now removed.

mirefek commented on graph2tac/tfgnn/dataset.py:
```diff
@@ -660,15 +644,14 @@ def _loader_to_proofstate_data(loader_data: tuple[LoaderProofstate, LoaderAction
         return (state_tuple, action_tuple, id)
 
-    def _proofstates(self) -> tf.data.Dataset:
+    def _proofstates(self, label : int = 0) -> tf.data.Dataset:
```
Maybe in case you don't do any split, so you don't want to think about them?

jasonrute replied:
This 0 should be TRAIN

mirefek commented on graph2tac/loader/data_server.py:
```diff
@@ -407,35 +462,34 @@ def _datapoint_text(self, i):
 
-    def data_train(self, shuffled: bool = False,  as_text: bool = False):
-        train_ids = self._select_data_points(0)
+    def get_datapoints(self, label : int, shuffled: bool = False, as_text: bool = False):
```
I would keep these further renaming ideas for the upcoming "dataset pipeline simplification", first I just wanted to reasonably add the splits

mirefek replied:
signature is fine

jasonrute replied:
I meant the output signature which is missing.  Although I realize it is more complicated than I thought.  I think it is something like `-> tuple[str, str] | tuple[LoaderProofstate, LoaderAction, int]`.

jasonrute replied:
Yes, it is fine to delay other simplications.

mirefek replied:
I added the signature (hopefully right)

mirefek commented:
I believe that with the last two commits, I addressed your comments, @LasseBlaauwbroek , @jasonrute . In particular more "static" typing, and consistency checks for the prefix labels. See if you have any further comments, or there is something I missed.

LasseBlaauwbroek commented on graph2tac/loader/data_server.py:
```diff
@@ -171,21 +172,95 @@ def cluster_to_graph(self, cluster):
+    def assign_data_server(self, data_server):
+
+        # conversion from graph indices to filenames
+        graphid_to_fname = {
+            file_data.graph : str(fname)
+            for fname, file_data in data_server._data.items()
+        }
+        assert set(graphid_to_fname.keys()) == set(range(len(data_server._data)))
+        graphid_to_fname = [
+            graphid_to_fname[i]
+            for i in range(len(data_server._data))
+        ]
```
Is all of this really needed? As far as I can tell, you never actually look something up in these maps. You only use the domain of the map.

mirefek replied:
It is a way to ensure that every graphid `0..len(data_server._data)-1` occurs there exactly once.

LasseBlaauwbroek replied:
You mean that you want to check that there is a bijection between filenames and graphid's? Okay, fine I guess, although if we want to test this, it might be better to do that centrally in pytact. (Although it seems slightly redundant, because in pytact essentially the opposite computation is happening to the one here...)

mirefek replied:
I have to do the inverse, and
```
        graphid_to_fname = {
            file_data.graph : str(fname)
            for fname, file_data in data_server._data.items()
        }
```
is in my view the most intuitive way of writing an inverse. Yes, I could keep it in a dictionary, and not convert to a list (but lists have some further advantages).

LasseBlaauwbroek replied:
Okay, this feels a little convoluted, but I see your point.

LasseBlaauwbroek reviewed (approved):
One final comment. Otherwise, assuming that this works, I propose to merge. There are still improvements to be made, but I agree with Mirek that this can be done as part of the bigger effort to simplify the data-pipeline.

mirefek commented:
The tests are not giving the preset values so far. I already had a discrepancy with Jason also before, now it seems completely different -- it can be due to a different way of splitting (so possibly differently shuffled) but maybe, I should look at it more carefully.

mirefek commented:
And @jasonrute also wanted something more with the tests -- add there an example of the prefix split, right?

jasonrute commented:
Yes, if you could add one test (under one of the datasets) that would be helpful.

jasonrute commented:
As for getting different numbers in the tests, since you changed the splits, I imagine it will change the numbers.  We are now splitting by lemma instead of proof step so we will have different train-valid split.  Also, when you removed the split from the Dataset pipeline, I'm sure it changed the randomness.  (We aren't using Jax, so we don't have a lot of control over the flow of the random number generator unfortunately.  One small change could change all the random numbers including the initial values of the network I assume.)

jasonrute commented on graph2tac/loader/data_server.py:
```diff
@@ -171,21 +172,95 @@ def cluster_to_graph(self, cluster):
+        ident_64 = np.array(d.node.identity, dtype = np.uint64).item()
+        return get_split_label(ident_64, self.proportions, self.random_seed)
+    def definition_cluster(self, d : list[Definition]) -> int:
+        return 0
```
Can we change this 0 to TRAIN?

jasonrute commented on graph2tac/loader/data_server.py:
```diff
@@ -171,21 +172,95 @@ def cluster_to_graph(self, cluster):
+        unused_prefixes = set().union(*[prefixes for label,prefixes in self.prefixes_per_label])
+        graphid_to_label = []
+        for fname in graphid_to_fname:
+            label = 0
```
Can we also change this 0 to TRAIN?

mirefek replied:
done

jasonrute replied:
Sorry, can you also look at all the other `0`'s and `1`'s in this data_server.py and train.py and see if they should be TRAIN or VALID?  Sorry to be so pedantic.

mirefek replied:
I think I did in the last commit, or have I missed some other?

mirefek commented:
Ah, makes sense, Fidel was not paying attention to emulating Vasily's split (which is used in the hash split).

jasonrute reviewed (approved):
LGTM.  Besides some of my pedantic comments, once we get the tests working, I think it should be good to merge.

jasonrute commented on graph2tac/tfgnn/dataset.py:
```diff
@@ -172,25 +158,23 @@ def graph_constants(self) -> GraphConstants:
             proofstate_datasets = {'train_proofstates': train_proofstates.batch(self.STATISTICS_BATCH_SIZE),
                                    'valid_proofstates': valid_proofstates.batch(self.STATISTICS_BATCH_SIZE)}
-            definition_dataset = self.definitions().batch(self.STATISTICS_BATCH_SIZE)
+            definition_dataset = self.definitions(0).batch(self.STATISTICS_BATCH_SIZE) # TODO: what are we using definitions for here?
```
This 0 should be TRAIN

jasonrute commented on graph2tac/tfgnn/train.py:
```diff
@@ -310,11 +306,9 @@ def run(self,
+        train_proofstates, valid_proofstates = self.dataset.proofstates(shuffle=True)
         if self.definition_task:
-            definitions = self.dataset.definitions(shuffle=True)
+            definitions = self.dataset.definitions(0, shuffle=False)
```
This 0 should be TRAIN

mirefek commented:
I updated the tests but since I checked them on my machine, they will likely not work for others...

LasseBlaauwbroek commented:
The tests pass for me :-)

mirefek commented:
Fine, I merge, further little bugs will get fixed later.

LasseBlaauwbroek commented:
Thanks Mirek!
